src/app/file/webp_format.cpp

```diff
@@ -0,0 +1,390 @@
+  FILE* fp = handle.get();
+
+  long len;
+  uint8_t* buf = NULL;
+
+  if (fseek(fp, 0, SEEK_END) != 0) {
+    fop_error(fop, "Error while getting WebP file size for %s\n", fop->filename.c_str());
+    return false;
+  }
+
+  len = ftell(fp);
+  rewind(fp);
+
+  if (len < 4) {
+    fop_error(fop, "%s is corrupt or not a WebP file\n", fop->filename.c_str());
+    return false;
+  }
+
+  buf = (uint8_t*) malloc(len);
+  if (!buf) {
+    fop_error(fop, "Error while allocating memory for %s\n", fop->filename.c_str());
+    return false;
+  }
+
+  if (!fread(buf, len, 1, fp)) {
+    fop_error(fop, "Error while writing to %s to memory\n", fop->filename.c_str());
+    free(buf);
+    return false;
+  }
+
+  WebPDecoderConfig config;
+  if (!WebPInitDecoderConfig(&config)) {
+    fop_error(fop, "LibWebP version mismatch %s\n", fop->filename.c_str());
+    free(buf);
+    return false;
+  }
+
+  if (WebPGetFeatures(buf, len, &config.input) != VP8_STATUS_OK) {
+    fop_error(fop, "Bad bitstream in %s\n", fop->filename.c_str());
+    free(buf);
+    return false;
+  }
+
+  fop->seq.has_alpha = config.input.has_alpha;
+
+  auto image = fop_sequence_image(fop, IMAGE_RGB, config.input.width, config.input.height);
+
+  config.output.colorspace = MODE_RGBA;
+  config.output.u.RGBA.rgba = (uint8_t*)image->getPixelAddress(0, 0);
+  config.output.u.RGBA.stride = config.input.width * sizeof(uint32_t);
+  config.output.u.RGBA.size = config.input.width * config.input.height * sizeof(uint32_t);
+  config.output.is_external_memory = 1;
+
+  WebPIDecoder* idec = WebPIDecode(NULL, 0, &config);
+  if (idec == NULL) {
+    fop_error(fop, "Error creating WebP decoder for %s\n", fop->filename.c_str());
+    free(buf);
+    return false;
+  }
+
+  auto bytes_remaining = len;
+  auto bytes_read = std::max(4l, len/100l);
+  auto data = buf;
+
+  while (bytes_remaining > 0) {
+    VP8StatusCode status = WebPIAppend(idec, data, bytes_read);
+    if (status == VP8_STATUS_OK || status == VP8_STATUS_SUSPENDED) {
+      bytes_remaining -= bytes_read;
+      data += bytes_read;
+      if (bytes_remaining < bytes_read) bytes_read = bytes_remaining;
+      fop_progress(fop, 1.0f - ((float)std::max(bytes_remaining, 0l)/(float)len));
+    } else {
+      fop_error(fop, "Error during decoding %s : %s\n", fop->filename.c_str(), WebPDecodingErrorMap.find(status)->second.c_str());
+      WebPIDelete(idec);
+      WebPFreeDecBuffer(&config.output);
+      free(buf);
+      return false;
+    }
+    if (fop_is_stop(fop))
+      break;
+  }
+
+  base::SharedPtr<WebPOptions> webPOptions = base::SharedPtr<WebPOptions>(new WebPOptions());
+  fop->seq.format_options = webPOptions;
+  webPOptions->lossless = std::min(config.input.format - 1, 1);
+
+  WebPIDelete(idec);
+  WebPFreeDecBuffer(&config.output);
+  return true;
+}
+
+#ifdef ENABLE_SAVE
+struct writerData {
+  FILE* fp;
+  FileOp* fop;
+};
+
+const std::pair<WebPEncodingError, std::string> enc_error_map_data[] = {
```

Comment:
I'm not sure about this array of pairs + a map. Is this just to show error messages? Could we have a function with a switch/case to convert `WebPEncodingError` codes to `const char*` strings (not even std::string)?